### DISCUSSION TITLE
Linkable DLL? What is WinGetUtil?

### DISCUSSION THREAD

**Discussion by josh-hemphill** [microsoft/winget-cli#2424] [Q&A] [answered]
Couldn't find anything concrete. I know PS piping is coming, but I want to pull WinGet info from an external application and was curious if there's a DLL I can link to instead of sending terminal commands to get the information.
I see WinGetUtil.dll mentioned in several places, but I can't quite tell from the cpp files what it covers, or where/how to find and link to it on Windows.

**Comment by florelis** [comment, marked as answer]
The Windows Package Manager exposes a COM API that you can use, but that is not exactly a DLL you can link. There is a [sample project](https://github.com/microsoft/winget-cli/tree/master/tools/SampleWinGetUWPCaller) for using it that you can reference. There's also a [spec for the API](https://github.com/microsoft/winget-cli/blob/master/doc/specs/%23888%20-%20Com%20Api.md), but surprisingly I could not find any actual documentation. @denelon, do we have docs for the COM API?

I'm not sure if WinGetUtil is something that we intend to be used externally. It is used by the validation service for packages submitted to winget-pkgs.

**Reply by josh-hemphill** [reply]
Do you know how to get the COM object? I tried loading the sample project and it appears as an unresolved reference, and I haven't been able to find it in any of the COM viewers.

**Reply by florelis** [reply]
Hard to know why it is an unresolved reference without knowing the exact error you get.

You need to reference the `Microsoft.Management.Deployment.winmd` from your project. The sample project uses a pre-compiled binary that is in this repo. You should use an actual released version but I'm actually not sure what is the right place to get it...

This should get better once the API itself gets to v1.0.
